posts/2021/06/pytorch-ignite-distributed.rst

```diff
@@ -0,0 +1,460 @@
+   .. code:: python
+
+      dist.init_process_group(backend, store=..., world_size=world_size, rank=rank)
+
+      rank = idist.get_rank()
+      ws = idist.get_world_size()
+      model = idist.auto_model(model)
+
+      dist.destroy_process_group()
+
+Running Distributed Code
+========================
+
+| PyTorch-Ignite's ``idist`` also unifies the distributed codes
+  launching method and makes the distributed configuration setup easier
+  with the
+  `ignite.distributed.launcher.Parallel (idist Parallel) <https://pytorch.org/ignite/distributed.html#ignite.distributed.launcher.Parallel>`__
+  context manager.
+| This context manager has the capability to either spawn
+  ``nproc_per_node`` (passed as a script argument) child processes and
+  initialize a processing group according to the provided backend or use
+  tools like ``torch.distributed.launch``, ``slurm``, ``horovodrun`` by
+  initializing the processing group given the ``backend`` argument only
+  in a general way.
+
+With ``torch.multiprocessing.spawn`` 
+------------------------------------
+
+In this case ``idist Parallel`` is using the native torch
+``torch.multiprocessing.spawn`` method under the hood in order to run
+the distributed configuration. Here ``nproc_per_node`` is passed as a
+spawn argument.
+
+-  Running multiple distributed configurations with one code. Source:
+   `ignite_idist.py <https://github.com/pytorch-ignite/idist-snippets/blob/master/ignite_idist.py>`__:
+
+.. code:: bash
+
+   # Running with gloo
+   python -u ignite_idist.py --nproc_per_node 2 --backend gloo
+
+   # Running with nccl
+   python -u ignite_idist.py --nproc_per_node 2 --backend nccl
+
+   # Running with horovod with gloo controller ( gloo or nccl support )
+   python -u ignite_idist.py --backend horovod --nproc_per_node 2
+
+   # Running on xla/tpu
+   python -u ignite_idist.py --backend xla-tpu --nproc_per_node 8 --batch_size 32
+
+With Distributed launchers
+--------------------------
+
+PyTorch-Ignite's ``idist Parallel`` context manager is also compatible
+with multiple distributed launchers.
+
+With torch.distributed.launch
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here we are using the ``torch.distributed.launch`` script in order to
+spawn the processes:
+
+.. code:: bash
+
+   python -m torch.distributed.launch --nproc_per_node 2 --use_env ignite_idist.py --backend gloo
+
+With horovodrun
+~~~~~~~~~~~~~~~~~~
+
+.. code:: bash
+
+   horovodrun -np 4 -H hostname1:2,hostname2:2 python ignite_idist.py --backend horovod
+
+.. note::
+   
+   In order to run this example and to avoid the installation procedure, you can pull one of PyTorch-Ignite's `docker image with pre-installed Horovod <https://github.com/pytorch/ignite/blob/master/docker/hvd/Dockerfile.hvd-base>`__. It will include Horovod with ``gloo`` controller and ``nccl`` support.
+
+   .. code:: bash
+
+      docker run --gpus all -it -v $PWD:/workspace/project --network=host --shm-size 16G pytorchignite/hvd-vision:latest /bin/bash
```

Comment:
maybe use a `\` to break this up? The line is twice as long as the rendered width of the post, so it's hard to read without scrolling.

Reply:
the `slurm` example below is nicely formatted

Reply:
Is this alright?
```      
docker run --gpus all -it -v $PWD:/workspace/project --network=host --shm-size 16G \
                   pytorchignite/hvd-vision:latest /bin/bash
```

Reply:
We can remove unimportant params:
```
docker run --gpus all -it $PWD:/project pytorchignite/hvd-vision:latest /bin/bash
```